tty/usbacm/usbacm.c

```diff
@@ -0,0 +1,565 @@
+				else {
+					msg.o.attr.err = -EINVAL;
+				}
+			default:
+				msg.o.io.err = -EINVAL;
+		}
+
+		msgRespond(usbacm_common.msgport, &msg, rid);
+	}
+
+	endthread();
+}
+
+
+static int _usbacm_handleInsertion(usb_devinfo_t *insertion)
+{
+	usbacm_dev_t *dev;
+	const usb_modeswitch_t *mode;
+	oid_t oid;
+
+	if ((mode = usb_modeswitchFind(insertion->descriptor.idVendor,
+			insertion->descriptor.idProduct,
+			modeswitch, sizeof(modeswitch) / sizeof(modeswitch[0]))) != NULL) {
+		return usb_modeswitchHandle(insertion, mode);
+	}
+
+	if ((dev = usbacm_devAlloc()) == NULL)
+		return -ENOMEM;
+
+	dev->instance = *insertion;
+
+	if ((dev->pipeCtrl = usb_open(insertion, usb_transfer_control, 0)) < 0) {
+		fprintf(stderr, "usbacm: Fail to open control pipe\n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	if (usb_setConfiguration(dev->pipeCtrl, 1) != 0) {
+		fprintf(stderr, "usbacm: Fail to set configuration\n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	if ((dev->pipeBulkIN = usb_open(insertion, usb_transfer_bulk, usb_dir_in)) < 0) {
+		fprintf(stderr, "usbacm: Fail to open bulk pipe\n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	if ((dev->pipeBulkOUT = usb_open(insertion, usb_transfer_bulk, usb_dir_out)) < 0) {
+		fprintf(stderr, "usbacm: Fail to open bulk pipe\n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	/* Interrupt pipe is optional */
+	dev->pipeIntIN = usb_open(insertion, usb_transfer_interrupt, usb_dir_in);
+
+	if (usbacm_init(dev) != EOK) {
+		free(dev);
+		fprintf(stderr, "usbacm: Init failed\n");
+		return -EINVAL;
+	}
+
+	oid.port = usbacm_common.msgport;
+	oid.id = dev->fileId;
+	if (create_dev(&oid, dev->path) != 0) {
+		free(dev);
+		fprintf(stderr, "usbacm: Can't create dev!\n");
+		return -EINVAL;
+	}
+
+	LIST_ADD(&usbacm_common.devices, dev);
+
+	fprintf(stderr, "usbacm: New device: %s\n", dev->path);
+
+	return 0;
+}
+
+
+static int _usbacm_threadJoin(usbacm_dev_t *dev)
+{
+	usbacm_dev_t *tmp;
+	int err;
+
+	while (!dev->thrFinished) {
+		if ((err = threadJoin(0)) < 0)
+			return err;
+
+		/* Find a device, whose thread has ended */
+		tmp = usbacm_common.devices;
+		do {
+			if (tmp->rxtid == err) {
+				tmp->thrFinished = 1;
+				break;
+			}
+			tmp = tmp->next;
+		} while (tmp != usbacm_common.devices);
+	}
+
+	return 0;
+}
+
+
+static int _usbacm_handleDeletion(usb_deletion_t *del)
+{
+	usbacm_dev_t *next, *dev = usbacm_common.devices;
+
+	if (dev == NULL)
+		return 0;
+
+	do {
+		next = dev->next;
+		if (dev->instance.bus == del->bus && dev->instance.dev == del->dev &&
+				dev->instance.interface == del->interface) {
+			fprintf(stderr, "usbacm: Device removed: %s\n", dev->path);
+			if (dev->fifo != NULL) {
+				_usbacm_threadJoin(dev);
+				/* Non-blocking device */
+				free(dev->fifo);
+				dev->fifo = NULL;
+				resourceDestroy(dev->fifoLock);
+				resourceDestroy(dev->fifoFull);
+				free(dev->stack);
+			}
+			remove(dev->path);
+			LIST_REMOVE(&usbacm_common.devices, dev);
+			free(dev);
+			if (dev == next)
+				break;
```

Comment:
Dangling pointer - don't ever use freed pointers. Do this check before free() and store the result